connectors/es/sink.py

```diff
@@ -356,33 +385,48 @@ async def _run(self):
             batch_num = 0
 
             while True:
+                force_flush = False
                 batch_num += 1
                 doc_size, doc = await self.fetch_doc()
-                if doc in (END_DOCS, EXTRACTOR_ERROR):
+                if isinstance(doc, ForceFlushSignal):
+                    self._logger.debug("Found ForceFlushSignal")
+                    force_flush = True
+                elif doc in (END_DOCS, EXTRACTOR_ERROR):
                     break
-                operation = doc["_op_type"]
-                doc_id = doc["_id"]
-                if not doc_id:
-                    self._logger.warning(f"Skip document {doc} as '_id' is missing.")
-                    continue
-                if operation == OP_DELETE:
-                    stats[operation][doc_id] = 0
                 else:
-                    # the doc_size also includes _op_type, _index and _id,
-                    # which we want to exclude when calculating the size.
-                    if overhead_size is None:
-                        overhead = {
-                            "_op_type": operation,
-                            "_index": doc["_index"],
-                            "_id": doc_id,
-                        }
-                        overhead_size = get_size(overhead)
-                    stats[operation][doc_id] = max(doc_size - overhead_size, 0)
-                self.counters.increment(operation, namespace=BULK_OPERATIONS)
-                batch.extend(self._bulk_op(doc, operation))
+                    operation = doc["_op_type"]
+                    doc_id = doc["_id"]
+                    if not doc_id:
+                        self._logger.warning(
+                            f"Skip document {doc} as '_id' is missing."
+                        )
+                        continue
+                    if operation == OP_DELETE:
+                        stats[operation][doc_id] = 0
+                    else:
+                        # the doc_size also includes _op_type, _index and _id,
+                        # which we want to exclude when calculating the size.
+                        if overhead_size is None:
+                            overhead = {
+                                "_op_type": operation,
+                                "_index": doc["_index"],
+                                "_id": doc_id,
+                            }
+                            overhead_size = get_size(overhead)
+                        stats[operation][doc_id] = max(doc_size - overhead_size, 0)
+                    self.counters.increment(operation, namespace=BULK_OPERATIONS)
+                    batch.extend(self._bulk_op(doc, operation))
+
+                    bulk_size += doc_size
+
+                if (
+                    len(batch) >= self.chunk_size
+                    or bulk_size > self.chunk_mem_size
+                    or force_flush
```

Comment:
Why do we need to add check for `force_flush` in here? The `force_flush = True` is set to True at every iteration of the loop, so the debug log will always get triggered if I understand correctly?

Reply:
`force_flush` is `True` only if a signal was received, which will happen really rarely - only if checkpoint is saved. Sorry I don't understand your comment fully :/